proposals/low-level-struct-improvements.md

```diff
@@ -0,0 +1,972 @@
+friction on `struct` members that don't participate in interface invocations. 
+
+One key compatibility scenario that we have to keep in mind when approaching 
+changes here is the following:
+
+```cs
+struct S1
+{
+    ref int GetValue() => ...
+}
+
+class Example
+{
+    ref int M()
+    {
+        // Okay: this is always allowed no matter how `local` is initialized
+        S1 local = default;
+        return local.GetValue();
+    }
+}
+```
+
+This works because the safety rules for `ref` return today do not take into 
+account the lifetime of `this` (because it can't return a `ref` to internal
+state). This means that `ref` returns from a `struct` can return outside the 
+enclosing method scope except in cases where there are `ref` parameters or a 
+`ref struct` which is not *safe-to-escape* outside the enclosing method scope. 
+Hence the solution here is not as easy as allowing `ref` return of fields in 
+non-interface methods.
+
+To remove this friction the language will provide the attribute `[RefEscapes]`.
+When this attribute is applied to an instance method, instance property or 
+instance accessor of a `struct` or `ref struct` then the `this` parameter will
+be considered *ref-safe-to-escape* outside the enclosing method.
+
+This allows for greater flexibility in `struct` definitions as they can begin 
+returning `ref` to their fields. That allows for types like `FrugalList<T>`:
+
+```cs
+struct FrugalList<T>
+{
+    private T _item0;
+    private T _item1;
+    private T _item2;
+
+    public int Count = 3;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            switch (index)
+            {
+                case 0: return ref _item1;
+                case 1: return ref _item2;
+                case 1: return ref _item3;
+                default: throw null;
+            }
+        }
+    }
+}
+```
+
+This will naturally, by the existing rules in the span safety spec, allow
+for returning transitive fields in addition to direct fields.
+
+```cs
+struct ListWithDefault<T>
+{
+    private FrugalList<T> _list;
+    private T _default;
+
+    public ref T this[int index]
+    {
+        [RefEscapes]
+        get
+        {
+            if (index >= _list.Count)
+            {
+                return ref _default;
+            }
+
+            return ref _list[index];
+        }
+    }
+}
+```
+
+Members which contain the `[RefEscapes]` attribute cannot be used to implement 
+interface members. This would hide the lifetime nature of the member at 
+the `interface` call site and would lead to incorrect lifetime calculations.
+
+To account for this change the "Parameters" section of the span safety document
+will be updated to include the following:
+
+- If the parameter is the `this` parameter of a `struct` type, it is 
+*ref-safe-to-escape* to the top scope of the enclosing method unless the 
+method is annotated with `[RefEscapes]` in which case it is *ref-safe-to-escape*
```

Comment:
I think the ref-escape of `[RefEscapes]` members will have to match val-escape of the container. 
If we are permitting returning refs to fields, we must ensure the containing variable is not dead by the time the ref is used.

I.E accessing ref-returning members on such structs would work as if accessing a field.

Reply:
Agree but I believe that falls out from the proposed rules. Essentially this is meant to change the *ref-safe-to-escape* scope for `this` when the `[RefEscapes]` attribute is applied. that will require an update to the `Parameters` section of the span safety rules. That should then feed into the existing `Field` rules for determining appropriate scopes.